magento/magento2
# Fixes: Bundle product with only one product in required bundle option not being added to cart with bundle option.

### Description (*)
Adding a bundle product to the cart from the product list widget doesn’t let the product be added properly with the bundle option. The bundle product options are not added within the parent product. It shows a warning message in minicart like ‘Please specify product option(s)’. It doesn’t fail in all the cases, it only fails in a certain case where a bundle product has been created with a required bundle option & the bundle option contains only one product with the ‘is_default’ checkbox selected.

Screenshot 1:
![image](https://github.com/magento/magento2/assets/106948523/dec6449b-882f-4239-bf5b-b43e178d6eb0)
Here you can see in the screenshot a bundle product has been created with a bundle option containing only one product in it & the option is required and the product is default select.

Screenshot 2:
![image](https://github.com/magento/magento2/assets/106948523/456174c7-2993-4449-a9eb-dbc2c34fdf56)
After creating a bundle product, a widget has been created with a type of Catalog Product List & a condition that matches the particular bundle product.

Screenshot 3:
![image](https://github.com/magento/magento2/assets/106948523/17382777-7351-4584-8064-384f59e263dd)
As we have created a widget, a bundle product is visible on the home page in the widget.

Screenshot 4:
![image](https://github.com/magento/magento2/assets/106948523/134bd1b7-cd63-4bf6-8286-459bd9ca3974)
When we click on the add-to-cart button, the product seems to be added to the cart & it doesn’t throw any error. And then if you check in the mini-cart, you will come to see a warning message like “Please specify product option(s).” 

The same thing works fine on a list page, but it fails while adding a product to the cart through the widget.

### Manual testing scenarios (*)
1. Create a bundle product that contains one or more bundle options & each bundle option is required and must contain only one product that is default selected. (see screenshot 1)
2. Create a widget with the type ‘Catalog Product List’ with a condition that includes the newly created bundle product in the widget. (see screenshot 2)
3. From the front end visit the widget & try to add the bundle product from the widget product list.(see screenshot 3)
4. After clicking Add to Cart observe a warning message in minicart. (see screenshot 4).

### Expected result (*)
The bundle product should be added to the cart with bundle options.

### Actual result (*)
The bundle product isn’t being added to the cart with bundle options.

### Additional Information (*)
File1: `Magento/CatalogWidget/view/frontend/templates/product/widget/content/grid.phtml`
File2: `Magento/Catalog/view/frontend/templates/product/list.phtml`

If we check in the above-given list.phtml, it gets the option by using the ViewModel & which works fine on the list page. (see screenshot 5).
Screenshot 5:
![image](https://github.com/magento/magento2/assets/106948523/fe8b7fb7-1db5-484b-b4b9-4334ea13e35d)

In the given `grid.phtml` file it doesn’t get the options data where the add-to-cart form is rendered for the particular product (see screenshot 6). As the options are not rendered, adding that product to the cart will only add the parent bundle product to the cart without any bundle options. So, we need to implement the same thing as `list.phtml` to get the data in `grid.phtml`.

Screenshot 6:
![image](https://github.com/magento/magento2/assets/106948523/8f286ae4-0398-405f-9691-3a61cc41a910)

### Solution (*)
We have to update the given phtml to get the bundle options in the add-to-cart form. We can get the bundle product options by using the `Magento\Catalog\ViewModel\Product\OptionsData` ViewModel. (see  screenshot 7)

A separate function has been created in the block file of the `grid.phtml` (`Magento\CatalogWidget\Block\Product\ProductsList`) to get the ViewModel that returns the object of `Magento\Catalog\ViewModel\Product\OptionsData`.

Screenshot 7:
![image](https://github.com/magento/magento2/assets/106948523/aadc093c-2cb4-4487-a71f-36004862828c)

### Contribution checklist (*)


### Resolved issues:
1. [x] resolves magento/magento2#38605: Fixes: Bundle product with only one product in required bundle option not being added to cart with bundle option.

### Related Pull Requests
https://github.com/magento-gl/magento2-page-builder/pull/84
